[1단계 - 장기] 라젤(문선영) 미션 제출합니다.

RaZel713 opened the pull request:
## 📖우테코 LEVEL1. 장기 미션 1단계

안녕하세요~ **오리** 리뷰어님!
우아한 테크 코스 7기 BE 라젤 입니다! ☺️

앞선 미션 중 다른 크루의 피드백을 통해 리뷰어님을 알게 되었는데, 생각해보면 좋은 점들에 대해 알려주시고, 굉장히 꼼꼼하게 봐주신 것을 보고 꼭 한번 만나 뵙고 싶었는데, 이렇게 이번 장기 미션을 통해 뵙게 되어 정말 기쁘고 반갑습니다! 
부족한 점들이 많지만, 잘 부탁드립니다!

---

### ✔️ 체크 리스트 ✔️

- [x]  미션의 필수 요구사항을 모두 구현했나요?
- [x]  Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x]  애플리케이션이 정상적으로 실행되나요?
- [ ]  [장기 미션 프롤로그](https://prolog.techcourse.co.kr/studylogs/4099)에 셀프 체크와 느낀 점을 작성했나요?


-  [📄 README.md](https://github.com/RaZel713/java-janggi/blob/step1/README.md)

---

### 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ]  1 (전혀 충족하지 못함)
- [x]  2
- [ ]  3 (보통)
- [ ]  4
- [ ]  5 (완벽하게 충족)
- 선택한 점수의 이유: 아직 개선 사항이 많이 남아 있을 것이라고 생각합니다!


### ⁉️ 궁금한 점 ⁉️

- Q1) Piece를 추상 클래스로 선언하여 작성하는 것이 좋은 선택 이였을 까요? 더 나은 선택이 있었을까요?

설계하는 과정에 있어, 
기물이 본인의 나라(즉, 플레이어의 진영)에 대하여 가지고 있어야 한다는 의견과
나라가 나라별 기물을 가지고 있어야 한다는 의견이 존재하였습니다.

이때, 저희는 각각의 경우에 따른 장단점에 대하여 고민해 기물이 본인의 나라를 가지고 있는 것으로 결정하였습니다.
따라서 상위 클래스는 ‘나라’라는 필드를 가져야 했고, 그로 인해 추상 클래스를 사용하게 되었습니다. (인터페이스는 동작만 가능하기 때문에)

하지만, 추상클래스를 사용했음에도 불구하고, 중복되는 부분들이 많다고 느낍니다.
일부 기물들의 특수한 규칙으로 인해 나머지 기물들에서 중복을 피하지 못한 부분들을 어떤 방식으로 해결하면 좋을까요?

인터페이스 (공통되는 동작) - 추상 클래스들 (동작이 일부 겹치는 기물들) - 구현체 (기물들)

이런 방식으로 작성하는 것이 더 나았을 까요? 


- Q2) 현재의 프로그램은, 

존재하는 Piece들을 관리하는 곳은 JanggiGame으로 해당 클래스에서는
Map<Dot, Piece> (기물이 존재하는 위치, 해당 기물)로 기물들을 관리하고 있습니다.

이때 말의 움직임을 판단하기 위해서는 Game에 소속되어있는 Piece가 Game에 존재하는 다른 말들에 대하여 알고 있어야합니다.

```
1. Piece의 이동경로에 다른 말이 있는지 판단하는데 사용된다.
2. Piece의 목적지에 자신과 같은 나라의 말이 있는지 판단하는데 사용된다.
```

Board 전체를 넘기는 것이 옳지 않다고 생각하여,
Piece에서 나라와 출발점, 목적지를 판단하여 해당 목적지로 가는 이동경로(List<Dot>)를 뽑아주고, 
해당 이동경로를 Board 에서 받아 해당 경로에 대한 Piece들을 찾아 다시 Piece에게 넘겨주어 해당 Piece들에 대해 2차 검증을 하는 방식으로 설계하였습니다.

이러한 방식으로 노출되는 정보를 줄일 수 있었지만, Board에서 관리되는 Piece가 Board에 존재하는 정보를 아는 것이 옳은 것인지 의문이 듭니다.

RaZel713 committed:
init: 초기화 테스트

RaZel713 committed:
init: 초기화 테스트

jinu0328 committed:
init: 초기화 테스트

jinu0328 committed:
docs: 1차 요구사항 업데이트

RaZel713 committed:
feat: x와 y좌표 범위를 제한하는 기능 구현

jinu0328 committed:
feat: 한나라와 초나라를 가지는 enum 생성

RaZel713 committed:
feat: 나라별 피스를 생성하는 기능 구현

jinu0328 committed:
feat: Piece의 타입을 지정하는 기능 구현

RaZel713 committed:
feat: 칸 생성 기능 구현

- 칸은 x좌표와 y좌표를 가진다
- 칸은 Piece를 가진다
- 칸에 Piece가 존재하지 않으면 Shadow Piece를 둔다

jinu0328 committed:
refactor: X좌표와 Y좌표 캐싱 기능 구현

RaZel713 committed:
refactor: LocationX, LocationY의 상수명 변경

RaZel713 committed:
feat: 장기판 생성 기능 구현

jinu0328 committed:
initial setting

RaZel713 committed:
docs: 기능 목록 작성

jinu0328 committed:
feat: x와 y 좌표를 가지는 점을 생성하는 기능 구현

- x와 y의 범위 유효성 검사 메서드 구현

RaZel713 committed:
feat: 주어진 범위에 맞는 점을 클래스 로드 시에 생성

jinu0328 committed:
feat: 나라를 가지는 기물을 생성하는 기능 구현

RaZel713 committed:
feat: 차의 출발점부터 목적지까지의 경로를 구하는 기능 구현

jinu0328 committed:
feat: 차의 이동 가능 여부 판단 기능 구현

RaZel713 committed:
feat: 기물을 추상클래스 Piece로 추상화

- 중복 메서드를 추상 클래스에서 관리

jinu0328 committed:
feat: 차의 경로 판단 시 도달할 수 없는 목적지의 경우 예외 발생 기능 구현

jinu0328 committed:
feat: 상의 목적지 경로 조회 및 이동 가능여부 판단 기능 구현

RaZel713 committed:
feat: 포의 목적지 경로 조회 및 이동 가능 여부 판단 기능 구현

- Piece의 추상 메서드 추가 및 수정

jinu0328 committed:
feat: 병의 목적지 경로 조회 및 이동 가능 여부 판단 기능 구현

RaZel713 committed:
feat: 마의 목적지 경로 조회 및 이동 가능 여부 판단 기능 구현

jinu0328 committed:
feat: 장, 사의 목적지 경로 조회 및 이동 가능 여부 판단 기능 구현

RaZel713 committed:
feat: 초기 기물 배치 전략 구현

jinu0328 committed:
fix: 한나라, 초나라 기물 배치 오류 수정

- 재할당 로직 제거후 putAll로 대체

jinu0328 committed:
feat: 기물을 포함한 장기판 출력 기능 구현

RaZel713 committed:
feat: 왕조와 출발점과 목적지를 기준으로 기물을 이동시키는 기능 구현

jinu0328 committed:
feat: 프로그램 실행 플로우 구현

RaZel713 committed:
fix: 포의 이동에 제한이 생겼을 경우 예외를 던지도록 수정

RaZel713 committed:
refactor: 패키지 분리

RaZel713 committed:
refactor: Dot의 일급 컬렉션 Board 구현

- 좌표 범위 제한 책임 Board로 위임

jinu0328 committed:
refactor: 한칸 움직임을 Dot 내부 메서드로 구현

RaZel713 committed:
refactor: 기물 움직임 메서드 분리

- 함수형 인터페이스 적용하여 가독성 향상

jinu0328 committed:
feat: 예외 발생 시 재시도 기능 구현

jinu0328 committed:
feat: 배치 전략 추가 구현 및 사용자 배치 전략 선택 기능 구현

RaZel713 committed:
refactor: 테스트 패키지 분리

jinu0328 committed:
docs: 입출력 결과 예시 추가

RaZel713 committed:
refactor: 미사용 구문 삭제

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Piece.java:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    public Dynasty getDynasty() {
```
> Piece를 추상 클래스로 선언하여 작성하는 것이 좋은 선택 이였을 까요? 더 나은 선택이 있었을까요?

어떤 것이 더 좋냐보다는 추상 클래스를 어떻게 사용했느냐가 더 중요할 것 같네요.

1. `protected`는 캡슐화를 지킨 것이 맞나요? `Piece`를 하나의 클래스로 본다면 상속받는 클래스가 그 정보를 가져가는것이 캡슐화인가요?
2. `validateSameDynasty`를 하위 클래스들이 public하게 사용하는데 상속을 하는데 하위에서 그걸 다시 호출하면 그건 상속의 적절한 사용법이 맞나요?
3. `validateSameDynasty`나 `getDynasty`를 override하여 변경이 되면 상속의 의미가 있나요?

RaZel713 replied:
1. 기물 클래스의 나라 필드를 private으로 변경하였습니다.
만약 해당 필드를 `protected`로 설정하면, 자식 클래스에서 부모 클래스의 필드에 접근할 수 있게 됩니다. 
이 경우, 자식 클래스에서 해당 필드를 직접 수정할 수 있게 되어 캡슐화 원칙을 지키지 못하게 됩니다. 
또한, `dynasty`는 `getDynasty()` 메서드를 통해 외부에서 해당 정보를 확인할 수 있으므로, 필드를 `protected`로 설정하는 것은 적합하지 않다고 판단하였습니다. 
따라서 해당 필드는 `private`으로 변경하여 캡슐화를 지키고, 필요한 경우 메서드를 통해 접근할 수 있도록 하였습니다.

2. 구현체를 만들고 공통되는 부분을 부모 클래스에 작성하는 방식으로 코드를 작성 하였는데, 그 과정에 있어 공통된 행위를 뽑아낸다는 것만 생각하여 작성한 것 같습니다.
주신 피드백을 보고 생각해보니, validateSameDynasty()는 경로에 존재하는 말을 판단하여 움직임 여부를 결정하는 메서드인 validateMove()에서만 활용되는 로직 임을 확인하여
validateMove() 안에 해당 로직을 작성하고, validateSameDynasty()를 제거하였습니다. 
그리고 구현체 내에서 validateMove()를 override 할 때, 
`super.validateMove(routesWithPiece, destinationPiece);`
부모 클래스의 로직을 호출한 후, 추가 사항을 판별하도록 변경하였습니다.

3. validateSameDynasty와 getDynasty는 변경이 필요로 하지 않는 부분이므로 final 키워드를 활용하여 override를 막았습니다. 추상 클래스를 작성하는 데 있어 final을 붙이지 않은 코드는 override 할 수 있다는 가능성을 생각하지 못하고 작성했던 것 같습니다! 알려주셔서 감사합니다.☺️

jinyoungchoi95 commented on src/main/java/janggiGame/JanggiGame.java:
```diff
@@ -0,0 +1,78 @@
+
+        Piece originPiece = pieces.get(origin);
+
+        List<Dot> route = originPiece.getRoute(origin, destination);
+
+        Map<Dot, Piece> routeWithPiece = getPiecesOn(route);
+
+        originPiece.validateMove(routeWithPiece, pieces.getOrDefault(destination, null));
```
> 이러한 방식으로 노출되는 정보를 줄일 수 있었지만, Board에서 관리되는 Piece가 Board에 존재하는 정보를 아는 것이 옳은 것인지 의문이 듭니다.

우선 `Piece`가 특정 지점에 도달하는 것을 확인하려면, `Piece`가 장기판의 정보를 받아 그 위치에 갈 수 있는지를 확인하는건 합당한 책임인 것 같아요.

다만 `Board`의 구조는 다른 이야기인 것 같네요. 해당 내용은 아래에서 코멘트 달아두겠습니다.

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Dynasty.java:
```diff
@@ -0,0 +1,16 @@
+package janggiGame.piece;
+
+public enum Dynasty {
+    HAN("한나라"),
+    CHO("초나라");
```
> 핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.

게임의 표기가 영어로 바뀌면 화면의 요구사항으로 인해 domain 코드에 변경점이 생기겠네요.

RaZel713 replied:
핵심 로직과 UI 로직을 구분하기 위해 

> 1. enum과 properties를 활용하여 UI 관련 내용을 properties에서 관리하게 구현
> 2. view에서 enum에 따라 이름을 반환해주는 로직 구현

이와 같은 방법을 생각해 보았습니다.

현재 dynasty, pieceType으로 한정되어있지만, 
표기 관련 사항이 늘어난다면, 2번 방법은 코드가 길어질 것이라고 생각하여 1번 방법을 활용하여 리팩터링하였습니다!

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Advisor.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Map;
+
+public class Advisor extends Piece {
+    public static final String NAME = "사";
```
위와 마찬가지입니다. 그리고 외부에 노출되지 않는 상수는 private이어야 할 것 같아요.

RaZel713 replied:
위의 comment에서 답변해두었습니다!

상수를 private 처리하는 것을 제대로 확인하지 못했습니다! 
좀 더 꼼꼼하게 확인하도록 하겠습니다! 감사합니다 😊

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Piece.java:
```diff
@@ -0,0 +1,32 @@
+
+    public abstract List<Dot> getRoute(Dot origin, Dot destination);
+
+    public abstract void validateRoute(int dx, int dy);
```
각 구현체에서 단순히 호출하는데 이걸 `Piece`의 스펙으로 둘 필요가 있나요?

`Piece`가 `validateRoute`라는 기능을 제공함으로써 자식 클래스가 아닌 다른 클래스와 상호 협력하는 기능이 있나요?

RaZel713 replied:
공통된 행위를 부모 클래스에서 작성해야 한다는 것만을 생각하여 
'해당 행위가 부모 클래스의 책임이 맞는가?'를 생각하지 않고 작성한 것 같습니다.

`validateRoute` 는 자식 클래스가 아닌 다른 클래스에서 상호 협력하는 기능이 없으므로,
`Piece`의 책임이 아닙니다.

따라서 `Piece`의 `validateRoute` 추상 메서드를 제거하고 각 구현체에서 private 메서드로 변경하여 작성하였습니다.
validateRoute는 각 자식 클래스에서만 필요한 로직으로, 부모 클래스의 행위가 아니므로 override 방식 대신 private 메서드로 변경하여 캡슐화하는 것이 맞는 것 같습니다.

해당 클래스의 책임에 대하여 생각하여 구현하도록 노력하겠습니다!
알려주셔서 감사합니다.😊

jinyoungchoi95 commented on src/main/java/janggiGame/board/Board.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Board {
```
`Board`의 객체로써의 역할은 무엇인가요? 다른 객체와 상호협력을 하고 있나요?

`Board`는 말그대로 판이라는 뜻인데 `Board`가 `Map<Dot, Piece>`를 관리하는 것이 더 적절한 책임 아닌가요?

RaZel713 replied:
처음에는 Dot 클래스에서 90개의 점을 생성하고, 입력 값에 맞춰 해당 점들을 반환하였습니다. 

하지만 Dot 클래스의 책임이 지나치게 커졌다고 느껴, 90개의 점을 관리하는 책임을 Dot에서 처리하는 것이 적합하지 않다고 판단하였습니다. 그래서 게임판을 관리하는 Board 클래스를 작성하여 90개의 Dot을 관리하도록 분리하였습니다.

> Board 클래스: 장기판을 구성하는 전체 점들을 관리
> JanggiGame 클래스: 존재하는 기물과 기물들의 좌표를 관리

그리고 Board로 해당 로직을 분리하는 과정에 있어 관련 로직들이 전역적으로 사용되기 때문에 메서드가 static 메서드로 작성되게 되었습니다.

하지만 리뷰어님께서 피드백 주신 내용을 바탕으로 고민하여 보니, 
 객체지향 설계에서 객체는 상태를 가지며 상태와 동작을 함께 관리해야 하는데, 
상태를 가지지 않는, 단순히 static 메서드만 제공하는 클래스는 객체라고 할 수 없는 것 같습니다. 

따라서 객체 지향적인 설계를 위해서는 90개의 점들을 다시 Dot 에서 관리하는 것이 맞다고 판단하여 Board 클래스를 삭제하고 90개의 점을 다시 Dot에서 관리하도록 변경하였습니다.

또한, JanggiGame 클래스의 명명을 Board라고 변경하였습니다.
JanggiGame이라는 이름은 게임의 흐름을 관리하는 역할처럼 보이기 때문에
클래스가 맡고 있는 책임인 존재하는 기물들을 관리하는 역할을 명확히 표현하는 데 적합하지 않다고 판단하였습니다.

이에 따라 JanggiGame 클래스 명을 Board로 변경하고, `Map<Dot, Piece> pieces`를 `survivedPieces`로 변경하여 생존한 기물들임을 보다 뚜렷하게 표현하도록 리팩터링하였습니다.

jinyoungchoi95 commented on src/main/java/janggiGame/board/Board.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    public static Dot findBy(Integer x, Integer y) {
```
단순히 static 메서드만 제공하는 클래스가 객체가 맞을지 고민해보세요.

jinyoungchoi95 commented on src/main/java/janggiGame/board/Dot.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    public Dot up() {
+        return Board.findBy(x, y + 1);
```
`Board`라는 것을 둘 필요 없이 9x9좌표계 자체를 `Dot`이 관리해야하지 않을까요? 그럼 이동할 수 있는 좌표계인지도 `Dot`이 처리가능해보이네요.

RaZel713 replied:
Board를 삭제하고 좌표계 자체를 Dot에서 관리하도록 리팩터링 하였습니다!

jinyoungchoi95 commented on src/main/java/janggiGame/board/Dot.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Objects;
+
+public class Dot {
+    private final Integer x;
+    private final Integer y;
```
`Integer`를 사용하는 이유가 있나요? wrapper class의 단점은 무엇일까요?

RaZel713 replied:
- row와 column 값은 한 번 설정되면 변경되지 않으며, 클래스가 불변 객체로 설계되었습니다. 불변 객체에서는 값이 변경되지 않으므로 기본 자료형인 int를 사용하는 것이 더 효율적입니다.
   - int는 기본 자료형으로 메모리를 적게 사용하고 연산 속도가 빠릅니다. 반면 Integer는 객체로 포장되어 있기 때문에 객체 생성 및 참조에서 추가 비용이 발생합니다.
   - Integer는 객체이므로 null 값을 허용하는데, 현재 코드에서는 row(전, x)와 column(전, y) 값이 항상 정의되어 있기 때문에 null 처리가 필요하지 않으며, int가 더 적합합니다.
   - int는 final을 통해 불변성을 보장할 수 있습니다.

따라서 row와 column은 int로 변경하는 것이 더 적합하다고 판단하여 int로 변경하였습니다!

(처음에는 불변성을 위해 Integer를 사용했지만, 피드백을 통해 int가 더 적합하다는 점을 알게 되었습니다. 알려주셔서 감사합니다! 🙏)

jinyoungchoi95 commented on src/main/java/janggiGame/arrangement/InnerElephantStrategy.java:
```diff
@@ -0,0 +1,45 @@
+        result.put(Board.findBy(1, 2), new Cannon(dynasty));
+        result.put(Board.findBy(7, 2), new Cannon(dynasty));
+
+        result.put(Board.findBy(0, 3), new Pawn(dynasty));
+        result.put(Board.findBy(2, 3), new Pawn(dynasty));
+        result.put(Board.findBy(4, 3), new Pawn(dynasty));
+        result.put(Board.findBy(6, 3), new Pawn(dynasty));
+        result.put(Board.findBy(8, 3), new Pawn(dynasty));
```
Strategy들에 공통되는 부분이 꽤 많아보이네요

RaZel713 replied:
현재 구현된 전략은 마와 상의 위치만 서로 변경할 수 있으며,  
나머지 기물(병, 장, 포, 차, 사)의 위치는 동일합니다.  

기존에는 ArrangementStrategy 인터페이스를 상속받아 각 장기 전략 구현체가 직접 모든 기물을 배치했지만,  
마와 상을 제외한 나머지 기물의 배치는 동일하므로 이를 분리하였습니다.  

이를 위해 HorseElephantChangeStrategy 클래스를 생성하여 마와 상을 제외한 기물을 배치하도록 하고,  
각 장기 전략 구현체는 해당 클래스를 부모 클래스로 두고 마와 상만 배치하도록 변경하였습니다.

jinyoungchoi95 commented on src/main/java/janggiGame/arrangement/ArrangementStrategy.java:
```diff
@@ -0,0 +1,10 @@
+import janggiGame.piece.Piece;
+import java.util.Map;
+
+public interface ArrangementStrategy {
```
하위클래스들이 실수할만한 부분이 많은 클래스들인 것 같은데 테스트는 필요없나요?

RaZel713 replied:
해당 코드는 특정 위치(점)에 특정 기물을 직접 배치하는 단순한 로직이므로, 테스트를 작성하지 않았습니다.  

테스트를 작성하는 과정에서, 해당 로직으로 생성된 Map에서 특정 위치에 특정 기물이 존재하는지 확인하는 방법이  
결국 해당 로직과 동일하다고 판단하였기 때문입니다.

하지만 이 로직 또한 도메인의 주요한 부분이라는 점에서 테스트의 필요성을 느껴 우선 해당 방식으로 테스트를 구현하였습니다.  

이에 대한 리뷰어님의 의견을 여쭈어 보고 싶습니다!

jinyoungchoi95 replied:
이미 테스트를 추가해주셨네요.

테스트하지 않아도되는 단순한 로직이라는 것은 주관적인 의견인 것 같아요. TDD를 떠올려본다면 본인의 테스트가 있어야 리팩터링을 진행하는데 믿고 수정할 포인트가 생길 것 같네요.

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Piece.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.List;
+import java.util.Map;
+
+public abstract class Piece {
```
추상 클래스의 테스트는 누가 해주어야 적절할까요?

RaZel713 replied:
추상 클래스는 직접 인스턴스를 만들 수 없기 때문에, 추상 클래스의 동작을 테스트 하려면 구체적인 구현체가 필요합니다.

추상 클래스를 구현한 구현체를 통해 추상 클래스를 테스트 할 수 있지만, 추후에 해당 구현체를 삭제하게 된다면 테스트 코드 또한 수정해야 하는 부담이 생깁니다.

따라서 실제 구현을 간소화하여 Fake 객체를 생성하고, 추상 클래스의 메서드가 예상대로 동작하는지 확인할 수 있게 하였습니다.

jinyoungchoi95 commented on src/test/java/janggiGame/board/DotTest.java:
```diff
@@ -0,0 +1,21 @@
+
+import static org.assertj.core.api.Assertions.assertThatCode;
+
+class DotTest {
```
up, down과 같은 기능들은 테스트가 필요 없을까요?

RaZel713 replied:
해당 기능들에 대한 테스트를 추가하였습니다!

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Pawn.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    @Override
+    public List<Dot> getRoute(Dot origin, Dot destination) {
+        int dx = origin.getDx(destination);
+        int dy = origin.getDy(destination);
+
+        validateRoute(dx, dy);
+
+        return List.of();
+    }
```
단순히 `Pawn`만 봤을 때 `getRoute`는 경로를 줄 수 있는 것처럼 메서드가 네이밍 되어있습니다. `Pawn`을 쓰려는 사람입장에서 `List.of()`라는 빈경로를 주는 것이라면 네이밍과 기능이 잘 매칭되고 있는 것이 맞나요?

RaZel713 replied:
**경로(Route)** 라는 단어는 일반적으로 출발점부터 도착점까지의 모든 지점을 포함하는 개념이므로,  
현재 로직이 출발점과 도착점을 제외한 나머지 지점만 반환하는 것과 의미가 맞지 않는 것 같습니다!

따라서 해당 메서드명을 출발점과 도착점을 제외한 경유하는 중간 점들을 반환한다는 의미를 지닌
`getIntermediatePoints()`로 변경하였습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 라젤 😄

장기 미션의 리뷰어를 맡은 오리입니다.

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

RaZel713 committed:
refactor: enum과 properties를 활용하여 출력을 위한 기물 명을 관리

RaZel713 committed:
refactor: 기물들의 불변 메서드를 오버라이딩할 수 없도록 제한

- validateSameDynasty()와 getDynasty() 메서드에 final 키워드를 추가하여 오버라이딩을 방지함

RaZel713 committed:
refactor: enum과 properties를 활용하여 출력을 위한 나라 명을 관리

RaZel713 committed:
refactor: 보드판을 구성하는 점들을 Dot에서 관리

RaZel713 committed:
refactor: Dot의 메서드, 상수 네이밍 변경

- x를 row로, y를 column으로 표현한다.
- 존재하는 90개의 점들에 한하여 값을 반환하고, 없으면 예외를 출력하므로 중복되는 유효성 검사 로직을 삭제한다.

RaZel713 committed:
refactor: validateMove에서 validateSameDynasty 로직 포함하여 삭제

- validateSameDynasty는 validateMove에서만 사용되므로 별도 메서드로 분리할 필요성이 없다.
- validateMove에 해당 로직을 통합하고 validateSameDynasty를 삭제하여 코드 간결화한다.

RaZel713 committed:
refactor: 불필요한 validateMove 오버라이드 삭제

- 부모 클래스의 validateMove를 그대로 사용하는 것이 가능하여 자식 클래스의 불필요한 오버라이드를 제거한다.

RaZel713 committed:
refactor: validateRoute 추상 메서드 제거 및 각 구현체에서 private 메서드로 변경

- validateRoute는 자식 클래스에서만 사용되며, 부모 클래스의 행위가 아니므로 부모 클래스에서 삭제한다.
- 각 자식 클래스에서만 필요한 로직이므로, 오버라이드 방식 대신 private 메서드로 변경하여 캡슐화한다.
- 불필요한 추상 메서드를 제거하여 부모 클래스의 책임을 명확하게 한다.

RaZel713 committed:
refactor: HorseElephantChangeStrategy 추가 및 장기 배치 전략 구조 변경

- 기존에는 인터페이스를 상속받아 각 전략에서 모든 기물을 배치했으나, 마와 상을 제외한 기물의 배치는 동일하므로 이를 공통 로직으로 분리한다.

RaZel713 committed:
refactor: 클래스 명 수정

RaZel713 committed:
test: 기물의 초기 배치 로직 테스트 구현

RaZel713 committed:
refactor: 기물의 추상 메서드 명 변경

- 기존 메서드명은 "경로"를 의미하는 getRoute였으나, 실제 반환값은 출발점과 도착점을 제외한 경유하는 점들만 포함한다.
- 따라서, 네이밍과 기능이 일치하도록 getIntermediatePoints()로 변경한다.

RaZel713 committed:
refactor: 테스트 이동

RaZel713 committed:
refactor: 기물 클래스의 나라 필드를 private으로 변경

- 캡슐화를 위해 기물 클래스의 `나라` 필드를 `private`으로 수정한다.

RaZel713 committed:
refactor: JanggiGame의 클래스명을 Board로 변경

RaZel713 committed:
test: 추상 클래스 Piece의 테스트 추가

RaZel713 committed:
refactor: 미사용 패키지 삭제

RaZel713 committed:
test: 방향에 따른 점 반환 테스트 추가

RaZel713 committed:
feat: 생성자에서 파라미터에 대한 NULL 검사 구현

jinyoungchoi95 commented on src/test/java/janggiGame/board/DotTest.java:
패키지가 테스트와 main 코드의 위치가 다르네요

RaZel713 replied:
test와 main 코드의 패키지가 일치하도록 변경하였습니다.

jinyoungchoi95 commented on src/main/java/janggiGame/Dot.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.List;
+import java.util.Objects;
+
+public class Dot {
```
"점"보다는 "위치"라는 클래스명이 더 적절하지 않을까요?

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Piece.java:
```diff
@@ -0,0 +1,28 @@
+        return dynasty;
+    }
+
+    public abstract List<Dot> getIntermediatePoints(Dot origin, Dot destination);
+
+    public void validateMove(Map<Dot, Piece> IntermediatePointsWithPiece, Piece destinationPiece) {
+        if (destinationPiece != null && destinationPiece.dynasty == this.dynasty) {
+            throw new UnsupportedOperationException("[ERROR] 같은 나라의 말은 공격할 수 없습니다.");
+        }
+    }
+
+    public abstract PieceType getType();
```
abstract와 일반 클래스의 정렬 순서가 어떻게 될까요?

RaZel713 replied:
abstract 메서드는 자식 클래스(상속받는 클래스)에서 반드시 구현해야하는 메서드 이므로 abstract를 먼저 정렬하여 
클래스가 어떤 동작을 반드시 구현해야하는 지 표현하는 것이 좋습니다.

현재의 코드 순서를 
**abstract 메서드 (필수 구현) -> override가 가능한 메서드 (추가 구현 가능) -> override가 불가능한 메서드** 
의 순서로 정렬하였습니다.

RaZel713 replied:
### 추상 메서드와 일반 메서드의 정렬 순서
✅ **추상 메서드를 먼저 배치한다**
	•	반드시 구현해야 하는 메서드가 먼저 보이므로, 클래스를 확장하는 개발자가 어떤 메서드를 반드시 구현해야 하는지 빠르게 파악할 수 있습니다.
	•	추상 메서드는 일반 메서드보다 상위 개념이므로  개념적 계층 구조를 명확히 할 수 있습니다.
	•	객체지향에서는 추상화(Abstraction)가 구체화(Concrete)보다 우선됩니다. 
	•	 API 문서 및 설계 관점에서 직관성을 높일 수 있습니다. 

> 예시로, Java의 AbstractList나 AbstractMap 등의 표준 라이브러리에서도 추상 메서드가 상단에 배치되어 있습니다.

✅ **추상 메서드를 마지막에 배치한다**
	•	구현 로직이 없는 메서드이므로, 구현된 내용이 있는 메서드(일반 메서드)가 먼저 나오는 것이 자연스럽다는 관점입니다.
	•	부모 클래스(추상 클래스)에서 중요한 것은 구현된 동작이 무엇인지이므로, 일반 메서드를 먼저 배치하면 실제 동작을 이해하는 데 도움이 될 수 있습니다. 
	•       기존 클래스를 수정하거나 디버깅할 때, 구현된 동작을 먼저 보고 흐름을 파악한 후, 마지막에 구현이 필요한 부분(추상 메서드)을 확인하는 것이 가독성이 더 좋을 수 있습니다. 
	•      메서드가 많아질 경우, 추상 메서드가 위쪽에 있으면 실제 구현된 동작을 찾기 어렵기 때문에 설계보다 사용성을 중심으로 했을 때 사용하는 코드 정렬 방식입니다.

> 예시로, Java 표준 라이브러리에서도 일부 클래스(예: AbstractCollection)는 일반 메서드를 먼저 배치하고, 추상 메서드를 마지막에 배치하였습니다.

이렇게 두가지 의견이 있는 것 같습니다! 

저는 추상 메서드를 우선 배치하는 것이 더 옳다고 생각하여 해당 방식을 적용 하였는데, 
해당 부분에 대하여 리뷰어님께 의견을 여쭙고 싶습니다!

_**실무에서는 어떤 방식으로 주로 사용되는지, 리뷰어님 개인의 의견으로는 어떻게 생각하시는 지 여쭈어봐도 될까요?**_

Java 표준 라이브러리에서 두 방법이 모두 존재하는 것으로 보아, 이는 클래스의 성격과 사용 패턴에 따라 다르게 적용될 수 있다고 생각이 듭니다! 
**_해당 부분은 컨벤션인 것 일까요..?_**

---

### 주요한 로직이 아닌 경우

`public abstract PieceType getType();`

해당 메서드는 필드의 값을 반환하는 getter는 아니지만, 구현체의 타입이 무엇인지 enum을 반환해주는 getter의 성향을 가지고 있습니다. 
보통 getter와 getter의 성향을 가진 메서드는 중요도가 떨어져 가장 마지막에 작성해야 한다고 생각하는데,
해당 메서드는 추상 메서드 이므로, 위에서 언급한 사항에 따르면 가장 윗 부분에 작성해야합니다. 
이러한 경우는 어떻게 하면 좋을까요..?

주요한 비즈니스 로직이 아니므로, 아래쪽에 작성해도 된다고 생각하였으나 우선 추상 메서드를 모아둔 윗부분의 가장 밑에 비치해두었습니다.

RaZel713 replied:
마지막으로,

✅ instanceof 사용을 피하기 위해 getType()을 활용하는 방식
1️⃣ **instanceof 사용을 피하려는 이유**
	•	instanceof는 OOP의 추상화 및 다형성 원칙을 깨트릴 가능성이 있습니다.
	•	새로운 구현체가 추가될 때마다 instanceof를 체크하는 코드도 변경되어야 하므로 유지보수성이 낮아집니다.
⸻

2️⃣ **적용한 방식**
	•	각 기물(Piece)의 클래스 이름 자체가 이미 해당 기물임을 나타내므로, 별도의 type 필드(예: PAWN, ROOK 등)를 가질 필요가 없습니다.
	•	따라서, getType()을 부모 추상 클래스에서 추상 메서드로 두고, 자식 클래스에서 Enum을 반환하도록 설계하였습니다.

해당방식에 대하여 어떻게 생각하시는지 의견을 여쭙고 싶습니다!

jinyoungchoi95 replied:
1. 실무에서 어떻게 쓴다거나하는 컨벤션은 따로 없는 것 같아요. 일관성있는 기준만 있으면 괜찮다고 생각합니다. 중요도보다는 abstract는 abstract끼리 구현메서드는 구현메서드끼리 두는 것이 더 일관성있겠네요.
2. 네 현재로도 충분해보이는데요. 반대로, `PieceType`이 `Piece`들을 가지는 구조를 만들면 type들을 별도로 선언하지 않아도 될 것 같네요.

jinyoungchoi95 commented on src/main/java/janggiGame/Board.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Map;
+
+public class Board {
+    private Map<Dot, Piece> survivedPieces = new HashMap<>();
```
```suggestion
    private final Map<Dot, Piece> survivedPieces = new HashMap<>();
```

jinyoungchoi95 commented on src/main/java/janggiGame/Board.java:
```diff
@@ -0,0 +1,75 @@
+
+        Map<Dot, Piece> intermediatePointsWithPiece = getPiecesOn(route);
+
+        originPiece.validateMove(intermediatePointsWithPiece, survivedPieces.getOrDefault(destination, null));
```
null을 사용하는 로직은 모두 제거하세요. 없는 경우엔 로직을 태우지 않으면되는데 null을 "없음"으로 볼 필요가 있을까요?

RaZel713 replied:
현재 코드는 목적지의 기물 여부를 검사하는 로직에 있어
빈 자리를 나타내는 표현이 필요하기 때문에,
null 을 사용하는 로직을 모두 없애고, "없음"을 뜻하는 `EmptyPiece`를 추가 하였습니다.

jinyoungchoi95 commented on src/main/java/janggiGame/Board.java:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private void validatePieceDynasty(Dynasty dynasty, Dot origin) {
+        if (survivedPieces.get(origin).getDynasty() != dynasty) {
```
getter를 사용하는 로직으로 보이네요. 객체가 메시지를 보내는 형태로 바꾸어주세요.

다른곳도 `getDynasty`를 사용하여 로직을 사용하는 곳들이 있으니 함께 수정해주세요.

RaZel713 replied:
`getDynasty()`을 제거하고 해당 로직을 사용하는 부분을

```
public final boolean hasDynasty(Dynasty dynasty) {
        return dynasty == this.dynasty;
}
```

위와 같이 `hasDynasty()`를 작성하여 
기물 객체에게 주어진 나라의 기물인지 여부를 받아 사용할 수 있게 변경하였습니다!

jinyoungchoi95 commented on src/main/java/janggiGame/Board.java:
```diff
@@ -0,0 +1,75 @@
+public class Board {
+    private Map<Dot, Piece> survivedPieces = new HashMap<>();
+
+    public void arrangeHanPieces(ArrangementStrategy strategy) {
+        Map<Dot, Piece> pieces = strategy.arrange(Dynasty.HAN);
+        Map<Dot, Piece> reversePieces = new HashMap<>();
+        pieces.keySet()
+                .forEach(dot -> reversePieces.put(dot.getReverse(), pieces.get(dot)));
+
+        this.survivedPieces.putAll(reversePieces);
+
+    }
+
+    public void arrangeChoPieces(ArrangementStrategy strategy) {
+        survivedPieces.putAll(strategy.arrange(Dynasty.CHO));
+    }
```
초기화 과정을 외부에 의존하지 않고, strategy 두개를 받아 생성하면 되지 않을까요? 둘을 호출하지 않은 `Board`를 믿고 사용할 수 없는 것 같네요.

RaZel713 replied:
프로그래밍 초반 설계에서는 Board가 90개의 위치를 관리하는 역할로 구상했지만,
현재는 **'생존한 기물들을 관리'** 하는 역할로 변경 되었음을 충분히 고려하지 못한 것 같습니다.

보드판이 존재하고(생성하고), 그 위에 기물을 배치(초기화) 한다는 현실적인 관점에서 생각하고 작성하다보니,
기물들의 초기화 과정이 Board의 책임에 맞지 않게 외부에 의존하게 되는 문제가 발생한 것 같습니다.

초반 배치 전략을 각각 입력 받고 해당 전략을 보드로 넘겨 보드에서 초기화하는 방식으로 변경하였습니다. 😁

jinyoungchoi95 commented on src/test/java/janggiGame/piece/AdvisorTest.java:
```diff
@@ -0,0 +1,41 @@
+        assertThat(actual).isEmpty();
+    }
+
+    @DisplayName("사는 목적지에 같은 나라의 기물이 존재한다면 이동할 수 없다")
```
이동할 수 있는 경우도 테스트하여야하지 않을까요?

RaZel713 replied:
각 기물의 테스트를 확인하여 이동할 수 있는 경우에 대한 테스트를 추가하였습니다.
또한, **예외 발생 테스트 → 정상 동작 테스트** 순서로 정렬하여 가독성을 높였습니다.

jinyoungchoi95 reviewed (changes requested):
안녕하세요 라젤 😄

몇가지 코멘트 남겨두었으니 확인부탁드려요 :)

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️

RaZel713 committed:
refactor: main 코드와 test 코드의 패키지 위치 일치

RaZel713 committed:
refactor: Board의 survivedPieces를 final로 변경하여 불변성 보장

RaZel713 committed:
refactor: Dot 클래스의 네이밍을 Position 으로 변경

RaZel713 committed:
refactor: 추상 클래스의 메서드 순서를 추상 메서드 -> 일반 메서드로 변경

RaZel713 committed:
refactor: Piece의 getter를 제거

- getDynasty() 대신 해당 나라의 기물인지 여부를 반환하게 하여 객체가 메시지를 보내주는 형식으로 변경

RaZel713 committed:
refactor: 보드 생성 시 두 나라의 전략을 받아 보드를 생성

RaZel713 committed:
test: 사의 테스트 추가

RaZel713 committed:
test: 포의 테스트 추가

RaZel713 committed:
test: 차의 테스트 추가

RaZel713 committed:
refactor: 상 테스트 수정

- 예외 발생 테스트 -> 정상 테스트 순으로 순서를 변경한다.
- 네이밍을 메서드에 맞춰 변경한다.

RaZel713 committed:
refactor: 마 테스트 수정

- 예외 발생 테스트 -> 정상 테스트 순으로 순서를 변경한다.
- 네이밍을 메서드에 맞춰 변경한다.

RaZel713 committed:
test: 장의 테스트 추가

RaZel713 committed:
test: 병의 테스트 추가

RaZel713 committed:
refactor: 빈 자리를 의미하는 기물 구현

- null을 사용하는 로직을 전부 EmptyPiece를 활용하여 변경한다.

RaZel713 committed:
refactor: 변수 명 변경

jinyoungchoi95 commented on src/main/java/janggiGame/Position.java:
```diff
@@ -0,0 +1,112 @@
+    private final int row;
+    private final int column;
+
+    public Position(final int row, final int column) {
```
캐싱한 값을 사용하는 것이 주 목적이라면 생성자는 막아주는 것이 좋겠네요.

jinyoungchoi95 commented on src/main/java/janggiGame/piece/Pawn.java:
```diff
@@ -0,0 +1,39 @@
+            throw new UnsupportedOperationException("[ERROR] 병이 이동할 수 있는 목적지가 아닙니다.");
+        }
+
+        if (hasDynasty(Dynasty.HAN) && dy > 0) {
+            throw new UnsupportedOperationException("[ERROR] 병은 뒤로 이동할 수 없습니다.");
+        }
+
+        if (hasDynasty(Dynasty.CHO) && dy < 0) {
+            throw new UnsupportedOperationException("[ERROR] 병은 뒤로 이동할 수 없습니다.");
+        }
```
한과 초에 대한 병 구현체를 별도로 두면 추상화되어있는 `Piece`를 구현체가 사용할 필요가 없어질 것 같네요.

jinyoungchoi95 reviewed (approved):
안녕하세요 라젤 😄

코멘트에 답변 남겨두었으니 다음 단계 진행하면서 함께 확인부탁드려요.

궁금하거나 고민이 되는 부분이 있으시다면 언제든 pr 코멘트 또는 dm으로 요청 부탁드립니다.
감사합니다 🙇‍♂️
